### PR TITLE
Add xxhash64 SparkSQL function

### PR DESCRIPTION
**Pull request by jinchengchenghh**

### CODE REVIEW

**Comment by mbasmanova** [comment]
@jinchengchenghh The are CI failures. Would you take a look? 

Please, update the PR title to use `Add xxx SparkSQL function` template.

**Review by mbasmanova — COMMENTED** [review]
@jinchengchenghh Some initial comments.

File: velox/functions/sparksql/Hash.cpp [hunk 1]
```diff
@@ -175,8 +339,23 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> hashSignatures() {
...
-  static const auto kHashFunction = std::make_shared<HashFunction>();
+  static const auto kHashFunction = std::make_shared<Murmur3HashFunction>();
   return kHashFunction;
 }
 
+std::vector<std::shared_ptr<exec::FunctionSignature>> xxhash64Signatures() {
+  return {exec::FunctionSignatureBuilder()
+              .returnType("bigint")
+              .argumentType("any")
+              .variableArity()
```
[the remark this answers is not in the document]
Comment (jinchengchenghh) [reply velox/functions/sparksql/Hash.cpp:350]: Yes, I have added StringInt32 test, is it enough?

File: velox/functions/sparksql/Hash.cpp [hunk 2]
```diff
@@ -35,130 +83,246 @@ namespace {
...
-
     context.ensureWritable(rows, INTEGER(), resultRef);
+    applyWithType<int32_t, Murmur3Hash>(
+        rows, args, outputType, context, resultRef);
+  }
+};
 
-    FlatVector<int32_t>& result = *resultRef->as<FlatVector<int32_t>>();
-    rows.applyToSelected([&](int row) { result.set(row, kSeed); });
+class XxHash64 final {
```
[the remark this answers is not in the document]
Comment (jinchengchenghh) [reply velox/functions/sparksql/Hash.cpp:178]: No, I compare the result of the velox/external/xxhash/xxhash.h, it is different with spark result

File: velox/functions/sparksql/tests/HashTest.cpp [hunk 3]
```diff
@@ -113,5 +113,102 @@ TEST_F(HashTest, Float) {
   EXPECT_EQ(hash<float>(-limits::infinity()), 427440766);
 }
 
+class XxHashTest : public SparkFunctionBaseTest {
```
[the remark this answers is not in the document]
Comment (jinchengchenghh) [reply velox/functions/sparksql/tests/HashTest.cpp:116]: sparksql and other project  https://learn.microsoft.com/en-us/dotnet/api/system.io.hashing.xxhash64?view=dotnet-plat-ext-7.0 naming it as `XxHash64`, Maybe XxHash64 is a common class name, Can I change it to `XxHash64Test`?

[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/functions/sparksql/tests/HashTest.cpp:116]: That works.

**Comment by jinchengchenghh** [comment]
Fixed according to all the comments

File: velox/common/base/BitUtil.h [hunk 4]
```diff
@@ -917,6 +917,13 @@ inline uint32_t rotateLeft(uint32_t a, int shift) {
 #endif
 }
 
+// Shift the bits of unsigned 64-bit integer a left by the number of
```
Comment (mbasmanova) [inline velox/common/base/BitUtil.h:920]: Use ///

Comment (mbasmanova) [inline velox/common/base/BitUtil.h:920]: > a left

to the left ?

File: velox/functions/sparksql/Hash.cpp [hunk 5]
```diff
@@ -35,130 +83,246 @@ namespace {
...
+class Murmur3HashFunction final : public exec::VectorFunction {
   bool isDefaultNullBehavior() const final {
     return false;
   }
 
   void apply(
       const SelectivityVector& rows,
       std::vector<VectorPtr>& args, // Not using const ref so we can reuse args
-      const TypePtr& /* outputType */,
+      const TypePtr& outputType,
```
Comment (mbasmanova) [inline velox/functions/sparksql/Hash.cpp:169]: Accidental change?

File: velox/functions/sparksql/Hash.cpp [hunk 6]
```diff
@@ -35,130 +83,246 @@ namespace {
...
+
+class XxHash64Function final : public exec::VectorFunction {
+  bool isDefaultNullBehavior() const final {
+    return false;
+  }
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args, // Not using const ref so we can reuse args
+      const TypePtr& outputType /* outputType */,
```
Comment (mbasmanova) [inline velox/functions/sparksql/Hash.cpp:320]: Remove commented out code.

File: velox/functions/sparksql/Hash.cpp [hunk 7]
```diff
@@ -175,8 +339,23 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> hashSignatures() {
...
     const std::vector<exec::VectorFunctionArg>& inputArgs) {
-  static const auto kHashFunction = std::make_shared<HashFunction>();
+  static const auto kHashFunction = std::make_shared<Murmur3HashFunction>();
   return kHashFunction;
 }
 
+std::vector<std::shared_ptr<exec::FunctionSignature>> xxhash64Signatures() {
+  return {exec::FunctionSignatureBuilder()
+              .returnType("bigint")
+              .argumentType("any")
```
Comment (mbasmanova) [inline velox/functions/sparksql/Hash.cpp:349]: This signature doesn't seem quite right. It suggests the function supports all types, but it looks like it doesn't support arrays, maps, structs, dates, timestamps, etc.

**Review by mbasmanova — COMMENTED** [review]
@jinchengchenghh Thank you for iterating on this PR. Looks good overall. Some comments below. Let's also run expressions fuzzer for an hour or so to make sure this function is tested well. See velox/expression/tests/SparkExpressionFuzzerTest.cpp .

File: velox/common/base/BitUtil.h [hunk 8]
```diff
@@ -917,6 +917,13 @@ inline uint32_t rotateLeft(uint32_t a, int shift) {
...
+// bits specified in shift, rotating the most-significant bit to the
+// least-significant bit location, and return the unsigned result.
+inline uint64_t rotateLeft64(uint64_t a, int32_t shift) {
```
[the remark this answers is not in the document]
Comment (jinchengchenghh) [reply velox/common/base/BitUtil.h:923]: No, I will change it to uint32_t

File: velox/functions/sparksql/Hash.cpp (hunk 7 above)
Comment (jinchengchenghh) [reply velox/functions/sparksql/Hash.cpp]: I try to specify the accurate function signature, but I don't know how to represent such case, do you have more suggested solution? Maybe like that https://github.com/facebookincubator/velox/commit/b5ab3be8960aaf98154c725b3f3c10ae9c8207a4 ?
```
Reason: Scalar function signature is not supported: xxhash64(VARCHAR, INTEGER).
```

Comment (jinchengchenghh) [reply velox/functions/sparksql/Hash.cpp:351]: I add the comment above function signature to describe supported types, does it work?

File: velox/functions/sparksql/Hash.h [hunk 9]
```diff
@@ -37,6 +37,20 @@ std::shared_ptr<exec::VectorFunction> makeHash(
     const std::string& name,
     const std::vector<exec::VectorFunctionArg>& inputArgs);
 
+// Supported types:
+//   - Bools
```
Comment (mbasmanova) [inline velox/functions/sparksql/Hash.h:41]: Bools -> Boolean

File: velox/functions/sparksql/Hash.h [hunk 10]
```diff
@@ -37,6 +37,20 @@ std::shared_ptr<exec::VectorFunction> makeHash(
...
+//   - Integer types (byte, short, int, long)
```
Comment (mbasmanova) [inline velox/functions/sparksql/Hash.h:42]: tinyint, smallint, integer, bigint

File: velox/functions/sparksql/Hash.h [hunk 11]
```diff
@@ -37,6 +37,20 @@ std::shared_ptr<exec::VectorFunction> makeHash(
...
+//   - String, Binary
```
Comment (mbasmanova) [inline velox/functions/sparksql/Hash.h:43]: Varchar, varbinary

File: velox/functions/sparksql/Hash.h [hunk 12]
```diff
@@ -37,6 +37,20 @@ std::shared_ptr<exec::VectorFunction> makeHash(
...
+//   - Float, Double
```
Comment (mbasmanova) [inline velox/functions/sparksql/Hash.h:44]: Real, double

File: velox/functions/sparksql/Hash.h [hunk 13]
```diff
@@ -37,6 +37,20 @@ std::shared_ptr<exec::VectorFunction> makeHash(
...
+//
+// Unsupported:
+//   - Decimal
+//   - Datetime
+//   - Structs, Arrays: hash the elements in order
```
Comment (mbasmanova) [inline velox/functions/sparksql/Hash.h:49]: > hash the elements in order

Is this a proposed solution? If so, put it as a TODO.

File: velox/functions/sparksql/Hash.h [hunk 14]
```diff
@@ -37,6 +37,20 @@ std::shared_ptr<exec::VectorFunction> makeHash(
...
+//   - Maps: iterate over map, hashing key then value. Since map ordering is
```
Comment (mbasmanova) [inline velox/functions/sparksql/Hash.h:50]: same

File: velox/functions/sparksql/tests/XxHash64Test.cpp [hunk 15]
```diff
@@ -28,6 +28,7 @@ class XxHash64Test : public SparkFunctionBaseTest {
   }
 };
 
+// expected result get from spark-sql=> select xxhash64("Spark")
```
Comment (mbasmanova) [inline velox/functions/sparksql/tests/XxHash64Test.cpp:31]: Comments must be full sentences, i.e. start with a capital letter and end with a period. Please, update throughout.

```
The expected result was obtained by running SELECT xxhash64("Spark") query using spark-sql CLI.
```

**Review by mbasmanova — COMMENTED** [review]
@jinchengchenghh Thank you for iterating on the PR. Looks great % a few nits.

**Comment by mbasmanova** [comment]
@jinchengchenghh Were you able to test this function using the fuzzer? Did it catch any bugs?

**Comment by mbasmanova** [comment]
@jinchengchenghh 

> I'm not sure if I test well, I receive 0 test.

Try `spark_expression_fuzzer_test  --logtostderr`

**Comment by jinchengchenghh** [comment]
root@sr249:/mnt/DP_disk1/code/velox/build/velox/expression/tests# ./spark_expression_fuzzer_test --logtostderr --duration_sec 600 --only xxhash64
xxhash64 is skipped

```
W1118 08:50:01.497020 3792411 ExpressionFuzzer.cpp:177] Skipping variadic function signature: xxhash64(any...) -> integer
```
```
I1118 08:50:01.496589 3792411 FuzzerRunner.h:120] Skipping special form: if
I1118 08:50:01.496654 3792411 FuzzerRunner.h:120] Skipping special form: coalesce
W1118 08:50:01.497020 3792411 ExpressionFuzzer.cpp:177] Skipping variadic function signature: xxhash64(any...) -> integer
I1118 08:50:01.497033 3792411 ExpressionFuzzer.cpp:335] Total candidate functions: 1 (1 signatures)
I1118 08:50:01.497040 3792411 ExpressionFuzzer.cpp:339] Functions with at least one supported signature: 0 (0.00%)
I1118 08:50:01.497048 3792411 ExpressionFuzzer.cpp:343] Functions with no supported signature: 1 (100.00%)
I1118 08:50:01.497056 3792411 ExpressionFuzzer.cpp:347] Supported function signatures: 0 (0.00%)
I1118 08:50:01.497058 3792411 ExpressionFuzzer.cpp:351] Unsupported function signatures: 1 (100.00%)
E1118 08:50:01.497076 3792411 Exceptions.h:68] Line: /mnt/DP_disk1/code/velox/velox/expression/tests/ExpressionFuzzer.cpp:644, Function:go, Expression: !signatures_.empty() No function signatures available., Source: RUNTIME, ErrorCode: INVALID_STATE
terminate called after throwing an instance of 'facebook::velox::VeloxRuntimeError'
  what():  Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: No function signatures available.
Retriable: False
Expression: !signatures_.empty()
Function: go
File: /mnt/DP_disk1/code/velox/velox/expression/tests/ExpressionFuzzer.cpp
Line: 644
Stack trace:
# 0
# 1
# 2
# 3
# 4
# 5
# 6
# 7
# 8
# 9
# 10

*** Aborted at 1668761401 (Unix time, try 'date -d @1668761401') ***
*** Signal 6 (SIGABRT) (0x39de1b) received by PID 3792411 (pthread TID 0x7f08bffc5bc0) (linux TID 3792411) (maybe from PID 3792411, UID 0) (code: -6), stack trace: ***
./spark_expression_fuzzer_test(_ZN5folly10symbolizer17getStackTraceSafeEPmm+0x80)[0x55e161afa0fd]
./spark_expression_fuzzer_test(_ZN5folly10symbolizer21SafeStackTracePrinter15printStackTraceEb+0x6a)[0x55e161c06cc8]
./spark_expression_fuzzer_test(+0x31791b3)[0x55e161c051b3]
./spark_expression_fuzzer_test(+0x3179294)[0x55e161c05294]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x1441f)[0x7f08c24ef41f]
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0xcb)[0x7f08c1fde00b]
/lib/x86_64-linux-gnu/libc.so.6(abort+0x12a)[0x7f08c1fbd858]
/lib/x86_64-linux-gnu/libstdc++.so.6(+0x9e910)[0x7f08c2397910]
/lib/x86_64-linux-gnu/libstdc++.so.6(+0xaa38b)[0x7f08c23a338b]
/lib/x86_64-linux-gnu/libstdc++.so.6(_ZSt9terminatev+0x16)[0x7f08c23a33f6]
/lib/x86_64-linux-gnu/libstdc++.so.6(__cxa_throw+0x48)[0x7f08c23a36a8]
./spark_expression_fuzzer_test(__cxa_throw+0x15c)[0x55e161af5cde]
./spark_expression_fuzzer_test(_ZN8facebook5velox6detail14veloxCheckFailINS0_17VeloxRuntimeErrorEPKcEEvRKNS1_18VeloxCheckFailArgsET0_+0x24a)[0x55e161a7e6db]
./spark_expression_fuzzer_test(_ZN8facebook5velox4test16ExpressionFuzzer2goEv+0x78)[0x55e1603d38f2]
./spark_expression_fuzzer_test(_ZN8facebook5velox4test16expressionFuzzerESt13unordered_mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt6vectorIPKNS0_4exec17FunctionSignatureESaISD_EESt4hashIS8_ESt8equal_toIS8_ESaISt4pairIKS8_SF_EEEm+0xab)[0x55e1603d41cb]
./spark_expression_fuzzer_test(_ZN12FuzzerRunner3runERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEmRKSt13unordered_setIS5_St4hashIS5_ESt8equal_toIS5_ESaIS5_EES7_+0x99)[0x55e1603bfd51]
./spark_expression_fuzzer_test(main+0x28a)[0x55e1603bcec6]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf2)[0x7f08c1fbf082]
./spark_expression_fuzzer_test(_start+0x2d)[0x55e1603ba93d]
(safe mode, symbolizer not available)
Aborted (core dumped)
```

**Comment by mbasmanova** [comment]
> Skipping variadic function signature: xxhash64(any...) -> integer

Try adding --enable_variadic_signatures flag.

**Comment by jinchengchenghh** [comment]
also skipped
```
root@sr249:/mnt/DP_disk1/code/velox/build/velox/expression/tests# ./spark_expression_fuzzer_test --logtostderr --duration_sec 600 --only xxhash64 --enable_variadic_signatures
I1118 09:20:21.569727 3830352 FuzzerRunner.h:120] Skipping special form: if
I1118 09:20:21.569795 3830352 FuzzerRunner.h:120] Skipping special form: coalesce
W1118 09:20:21.570205 3830352 ExpressionFuzzer.cpp:199] Skipping unsupported signature with generic: xxhash64(any...) -> integer
I1118 09:20:21.570225 3830352 ExpressionFuzzer.cpp:335] Total candidate functions: 1 (1 signatures)
I1118 09:20:21.570231 3830352 ExpressionFuzzer.cpp:339] Functions with at least one supported signature: 0 (0.00%)
I1118 09:20:21.570245 3830352 ExpressionFuzzer.cpp:343] Functions with no supported signature: 1 (100.00%)
I1118 09:20:21.570251 3830352 ExpressionFuzzer.cpp:347] Supported function signatures: 0 (0.00%)
I1118 09:20:21.570256 3830352 ExpressionFuzzer.cpp:351] Unsupported function signatures: 1 (100.00%)
```

**Comment by mbasmanova** [comment]
> Skipping unsupported signature with generic: xxhash64(any...) -> integer

Oh... @kagamiori Wei, do you know if this is expected?

**Comment by mbasmanova** [comment]
@jinchengchenghh Can you try adding --velox_fuzzer_enable_complex_types flag ?

**Comment by jinchengchenghh** [comment]
same reason to skip
```
root@sr249:/mnt/DP_disk1/code/velox/build/velox/expression/tests# ./spark_expression_fuzzer_test --logtostderr --duration_sec 600 --only xxhash64 --enable_variadic_signatures --velox_fuzzer_enable_complex_types
I1118 09:24:53.559481 3835437 FuzzerRunner.h:120] Skipping special form: if
I1118 09:24:53.559547 3835437 FuzzerRunner.h:120] Skipping special form: coalesce
W1118 09:24:53.559940 3835437 ExpressionFuzzer.cpp:199] Skipping unsupported signature with generic: xxhash64(any...) -> integer

```

**Comment by mbasmanova** [comment]
@jinchengchenghh I guess Fuzzer doesn't support this type of function just yet. Thanks for trying.

**Comment by kagamiori** [comment]
> > Skipping unsupported signature with generic: xxhash64(any...) -> integer
> 
> Oh... @kagamiori Wei, do you know if this is expected?

We only support "any" if the function signature also contain type variables right now. We'll add this support for all function signatures (https://github.com/facebookincubator/velox/issues/3310).

**Comment by jinchengchenghh** [comment]
Can this one be merge? @mbasmanova @kagamiori

**Comment by mbasmanova** [comment]
@jinchengchenghh There is a failure in format-check CI job. Run make format-fix and update the PR.

**Comment by mbasmanova** [comment]
@jinchengchenghh I'm seeing some failures with UBSAN:

```
[----------] 1 test from XxHash64Test
[ RUN      ] XxHash64Test.float

stderr:
velox/functions/sparksql/Hash.cpp:191:35: runtime error: signed integer overflow: 1065353216 * -7046029288634856825 cannot be represented in type 'long'
    #0 0x7fa2bda841ef in facebook::velox::functions::sparksql::(anonymous namespace)::XxHash64::hashInt32(int, long) velox/functions/sparksql/Hash.cpp:191
    #1 0x7fa2bda8faaf in facebook::velox::functions::sparksql::(anonymous namespace)::XxHash64::hashFloat(float, long) velox/functions/sparksql/Hash.cpp:206
    #2 0x7fa2bda8f64b in void facebook::velox::functions::sparksql::(anonymous namespace)::applyWithType<long, facebook::velox::functions::sparksql::(anonymous namespace)::XxHash64>(facebook::velox::SelectivityVector const&, std::vector<std::shared_ptr<facebook::velox::BaseVector>, std::allocator<std::shared_ptr<facebook::velox::BaseVector> > >&, facebook::velox::exec::EvalCtx&, std::shared_ptr<facebook::velox::BaseVector>&)::'lambda7'(int)::operator()(int) const velox/functions/sparksql/Hash.cpp:70

SUMMARY: UndefinedBehaviorSanitizer: signed-integer-overflow velox/functions/sparksql/Hash.cpp:191:35 in 
```

**Comment by mbasmanova** [comment]
Also, there is a lint error:

Screen Shot 2022-11-21 at 8 10 51 PM

**Comment by jinchengchenghh** [comment]
I use `reinterprete_cast` to fix this proplem `signed integer overflow`.
Does this work? @mbasmanova

**Comment by mbasmanova** [comment]
@jinchengchenghh I still get an error:

```
stderr:
velox/functions/sparksql/Hash.cpp:253:38: runtime error: signed integer overflow: -7046029288634856783 + -4417276706812531889 cannot be represented in type 'long'
    #0 0x7f0193a8bd8e in facebook::velox::functions::sparksql::(anonymous namespace)::XxHash64::hashBytesByWords(facebook::velox::StringView const&, long) velox/functions/sparksql/Hash.cpp:253
    #1 0x7f0193a8b825 in facebook::velox::functions::sparksql::(anonymous namespace)::XxHash64::hashBytes(facebook::velox::StringView const&, long) velox/functions/sparksql/Hash.cpp:219
```

**Pull request closed without merging**